core/testcontainers/core/container.py

```diff
@@ -83,14 +86,63 @@ def maybe_emulate_amd64(self) -> Self:
             return self.with_kwargs(platform="linux/amd64")
         return self
 
+    def depends_on(self, dependencies: Union["DockerContainer", list["DockerContainer"]]) -> "DockerContainer":
+        """
+        Specify dependencies for this container.
+
+        Args:
+            dependencies (Union[DockerContainer, list[DockerContainer]]): One or multiple Docker container instances
+            this container depends on.
+
+        Returns:
+            DockerContainer: The current instance, for chaining.
+        """
+        if isinstance(dependencies, DockerContainer):
+            self._dependencies.append(dependencies)
+        elif isinstance(dependencies, list):
+            self._dependencies.extend(dependencies)
+        else:
+            raise TypeError("dependencies must be a DockerContainer or list of DockerContainer instances")
+        return self
+
+    def _start_dependencies(self) -> bool:
+        """
+        Start all dependencies recursively, ensuring each dependency's dependencies are also resolved.
+        If a dependency fails to start, stop all previously started dependencies and raise the exception.
+        """
+        started_dependencies = []
+        for dependency in self._dependencies:
+            if not dependency._container:
+                try:
+                    dependency._start_dependencies()
+                    dependency.start()
+                    started_dependencies.append(dependency)
+
+                    if not dependency.wait_until_running():
+                        raise ContainerStartException(f"Dependency {dependency.image} did not reach 'running' state.")
+
+                except Exception as e:
+                    # Clean up all dependencies started before the failure
+                    for dep in started_dependencies:
```

Comment:
this doesn't seem like it work as the recursive call would create its own started_dependencies?

Reply:
@alexanderankin I have made the needed changes such that the _start_dependencies function passes the started_dependencies across the recursive function calls to maintain a consistent state. I have also added a new check for circular dependencies which is called while adding dependencies in depends_on function. The tests have also been updated to be more comprehensive. I will be happy to make changes in the future if any of the core parts change just tag me in that case.